conductor-modules/viewpager2/src/main/java/com/bluelinelabs/conductor/viewpager/RouterStateAdapter.kt

```diff
@@ -0,0 +1,190 @@
+      throw IllegalStateException("Expected ViewPager2 instance. Got: $parent")
+    }
+    return parent
+  }
+
+  override fun onAttachedToRecyclerView(recyclerView: RecyclerView) {
+    val viewPager = inferViewPager(recyclerView)
+    primaryItemCallback = PrimaryItemCallback().also {
+      viewPager.registerOnPageChangeCallback(it)
+    }
+  }
+
+  override fun onDetachedFromRecyclerView(recyclerView: RecyclerView) {
+    val viewPager = inferViewPager(recyclerView)
+    primaryItemCallback?.let {
+      viewPager.unregisterOnPageChangeCallback(it)
+    }
+    primaryItemCallback = null
+  }
+
+  override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RouterViewHolder {
+    return RouterViewHolder.create(parent)
+  }
+
+  override fun onBindViewHolder(holder: RouterViewHolder, position: Int) {
+    val itemId = getItemId(position)
+    val router = host.getChildRouter(holder.container, "$itemId")
+    holder.currentRouter = router
+    holder.currentItemId = itemId
+    holder.currentItemPosition = position
+
+    if (!router.hasRootController()) {
+      val routerSavedState = savedPages[position.toLong()]
+      if (routerSavedState != null) {
+        router.restoreInstanceState(routerSavedState)
+        savedPages.remove(itemId)
+        savedPageHistory.remove(itemId)
+      }
+    }
+
+    router.rebindIfNeeded()
+    configureRouter(router, position)
+
+    if (position != currentPrimaryRouterPosition) {
+      for (transaction in router.backstack) {
+        transaction.controller.setOptionsMenuHidden(true)
+      }
+    }
+
+    visibleRouters.put(position, router)
+  }
+
+  override fun onViewRecycled(holder: RouterViewHolder) {
+    holder.currentRouter?.let { router ->
+      val savedState = Bundle()
+      router.saveInstanceState(savedState)
+      savedPages.put(holder.currentItemId, savedState)
+
+      savedPageHistory.remove(holder.currentItemId)
+      savedPageHistory.add(holder.currentItemId)
+
+      ensurePagesSaved()
+
+      host.removeChildRouter(router)
+
+      visibleRouters.remove(holder.currentItemPosition)
+    }
+  }
+
+  override fun onFailedToRecycleView(holder: RouterViewHolder): Boolean {
+    return true
+  }
+
+  override fun getItemId(position: Int): Long {
+    return position.toLong()
+  }
+
+  override fun setHasStableIds(hasStableIds: Boolean) {
+    throw UnsupportedOperationException("Stable Ids are required for the adapter to function properly")
+  }
+
+  override fun saveState(): Parcelable {
+    return SavedState(
+      savedPagesKeys = (0..savedPages.size()).map { savedPages.keyAt(it) },
+      savedPagesValues = (0..savedPages.size()).map { savedPages.valueAt(it) },
+      savedPageHistory = savedPageHistory,
+      maxPagesToStateSave = maxPagesToStateSave
+    )
+  }
+
+  override fun restoreState(state: Parcelable) {
+    if (state !is SavedState) return
+
+    savedPages = LongSparseArray()
+    (0..state.savedPagesKeys.size).forEach { index ->
```

Comment:
0 unitl ..., this is one index too far.

Reply:
Alternatively, `0..state.savedPagesKeys.lastIndex`

Reply:
incidces.forEach :)

Reply:
I tend to not use these android collections at all these days because you lose the kotlin collection stuff and it doesn't matter here anyways (this think holding multiple views, here a few ints are just nuts)

Reply:
oof…